cmd/abapEnvironmentAssembly.go

```diff
@@ -0,0 +1,598 @@
+	c.Stdout(log.Writer())
+	c.Stderr(log.Writer())
+
+	// for http calls import  piperhttp "github.com/SAP/jenkins-library/pkg/http"
+	// and use a  &piperhttp.Client{} in a custom system
+	// Example: step checkmarxExecuteScan.go
+
+	// error situations should stop execution through log.Entry().Fatal() call which leads to an os.Exit(1) in the end
+	err := runAbapEnvironmentAssembly(&config, telemetryData, &c, commonPipelineEnvironment)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runAbapEnvironmentAssembly(config *abapEnvironmentAssemblyOptions, telemetryData *telemetry.CustomData, command command.ExecRunner, commonPipelineEnvironment *abapEnvironmentAssemblyCommonPipelineEnvironment) error {
+	return runAssembly(config)
+}
+
+// *******************************************************************************************************************************
+// ********************************************************** Step logic *********************************************************
+// *******************************************************************************************************************************
+func runAssembly(config *abapEnvironmentAssemblyOptions) error {
+	conn := new(connector)
+	err := conn.init(*config, &piperhttp.Client{})
+	if err != nil {
+		return err
+	}
+	assemblyBuild := build{
+		connector: *conn,
+	}
+
+	valuesInput := values{
+		Values: []value{
+			{
+				ValueID: "SWC",
+				Value:   config.SWC,
+			},
+			{
+				ValueID: "CVERS",
+				Value:   config.SWC + "." + config.SWCRelease + "." + config.SpsLevel,
+			},
+			{
+				ValueID: "NAMESPACE",
+				Value:   config.Namespace,
+			},
+			{
+				ValueID: "PACKAGE_NAME_" + config.PackageType,
+				Value:   config.PackageName,
+			},
+		},
+	}
+	if config.PreviousDeliveryCommit != "" {
+		valuesInput.Values = append(valuesInput.Values,
+			value{ValueID: "PREVIOUS_DELIVERY_COMMIT",
+				Value: config.PreviousDeliveryCommit})
+	}
+	phase := "BUILD_" + config.PackageType
+	err = assemblyBuild.startPollLog(phase, valuesInput, time.Duration(config.MaxRuntimeInMinutes), 60)
+	if err != nil {
+		return err
+	}
+	resultName := "SAR_XML"
+	resultSARXML, err := assemblyBuild.getResult(resultName)
+	if err != nil {
+		return err
+	}
+	envPath := filepath.Join(GeneralConfig.EnvRootPath, "commonPipelineEnvironment")
+	downloadPath := filepath.Join(envPath, path.Base(resultName))
+	err = resultSARXML.download(downloadPath)
+	return err
+}
+
+// *******************************************************************************************************************************
+// ************************************************************ REUSE ************************************************************
+// *******************************************************************************************************************************
+type resultState string
+type runState string
+type msgty string
+
+const (
```

Comment:
Maybe that's just a personal preference: I find it easier to navigate the file when the structs etc. are at the bottom of the file. Usually, I read the functions more often than the structs.

Reply:
I did this, however added one more struct(buildWithRepository) which is really step specific (and not part of the REUSE part) therefore I kept this on the upper part.